gitpod-io/gitpod-vscode-desktop
# Implement local ssh daemon

## Description

Implement local ssh daemon

## Related Issue(s)

Related https://github.com/gitpod-io/gitpod/issues/16836
Related https://linear.app/gitpod/issue/IDE-29/spike-compare-localssh-%E2%9A%94%EF%B8%8F-directssh-localapp

## Analysis Path [^1]
- Event vscode_desktop_ssh, with property `kind: local-ssh`
- With funnel: `connecting` -> `? local-ssh-failure` -> connected (`got_destination`)

[^1]: https://github.com/gitpod-io/gitpod-vscode-desktop/pull/36#discussion_r1152898622

## Error Reporting [^1]

Images below are local testing result, it will send to GCP in production
|String|Error|
|:-:|:-:|
|<img width="1259" alt="image" src="https://user-images.githubusercontent.com/20944364/232742671-d22a2e5a-518c-406d-8ca8-771743e308f6.png">|<img width="1135" alt="image" src="https://user-images.githubusercontent.com/20944364/232743619-b70a172c-d6e0-4bc5-9048-f4ead9f94f30.png">|


## How to test


Should test with platform `stable`, `insiders` and os `linux` `macOS` `Windows` (mostly test with file system)

- Clone this repo and switch to PR's branch
- Exec commands below to install extension locally
```
yarn
yarn local-test
```
- Restart VSCode to make sure daemon is up-to-date (need 10 seconds to exit) [^2], commands below can help test [^3]
- Start workspace in Gitpod and Open in VSCode
- It should able to start without password required
- Stop workspace, and use VSCode Desktop `recent` (ctrl + R) to access the stopped workspace
- It should ask you if you want to restart it
- Run command `Gitpod: Export all Logs` should include local ssh loggings
  - [x] Insiders

### ‼️  Major Test Points ‼️
- [x] ‼️ Local SSH can be controlled by  feature flag https://app.configcat.com/08da1258-64fb-4a8e-8a1e-51de773884f6/08da1258-6541-4fc7-8b61-c8b47f82f3a0/08da1258-6512-4ec0-80a3-3f6aa301f853?settingId=75503
- [ ] Throw error inside localssh (i.e. reject [here](https://github.com/gitpod-io/gitpod-vscode-desktop/blob/f2d52c8e5ef97ef5d7c58d4168413320bcd825c1/src/local-ssh/sshTunnel.ts#L59)) errors should reported to segment https://app.segment.com/gitpod/sources/ide_team_playground/debugger
- [ ] Should send event to segment https://app.segment.com/gitpod/sources/ide_team_playground/debugger with `kind = local-ssh`
- [x] ‼️ It can connect with LocalSSH
  - [x] Insiders

[^2]: We need 10 second to avoid daemon exists too quickly with only one window trying to reload
[^3]: watch daemon process `watch -n 1 "ps aux | grep '[d]aemon.js'"`  and tail daemon loggings `tail -f $TMPDIR/gitpod-vscode-daemon.log | grep <pid>`
